### ISSUE TITLE
HWLB overrides DC-based replica preference

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#9077]
HWLB (Heat Weighted Load Balancing, see [this blog post](https://www.scylladb.com/2017/09/21/scylla-heat-weighted-load-balancing/), may change the replicas that a read request uses, overriding the snitch's preferred replicas. As can be seen in `db/consistency_level.cc`, the HWLB code is only avoided if all replicas' hit ratios are above 0.95 (`max_hit_rate`), or if they are lower but close to each other within 0.01 (`ht_max - ht_min > 0.01`). But in other cases, the HWLB code may be used.

When HWLB is used, it considers all replicas in the cluster, ignoring their placement into DCs. This can cause it to pick a node in a remote DC with a hot cache over a node in the local DC with a cold cache.

The thing is, Scylla usually prefers (this preference is expressed via a _snitch_) to send requests to nodes in the local DC, because a request sent to a remote DC will experience high cross-DC latencies. So if HWLB believes that by sending a request to a hot-cache node on a remote DC it will reduce latency, it is mistaken. Especially in a modern setup where disks have very low latency but cross-DC traffic has high latency.

It's important to note that we **don't** have this problem with reads with "local" consistency levels (`LOCAL_ONE`, `LOCAL_QUORUM`) - importantly, those reads will not violate the LOCAL requirement and not go to remote DCs: When the consistency level is local, the code (https://github.com/scylladb/scylla/blob/scylla-4.5.rc5/db/consistency_level.cc#L176) begins by keeping only the "local" (current DC) nodes in the list of live nodes - so HWLB cannot choose nodes in other DCs. 

We should consider changing HWLB to be aware of DCs. Perhaps if the selected replicas are all from one DC, don't consider replicas in other DCs.

CC @gleb-cloudius @vladzcloudius

**Comment by nyh** [comment]
@mykaul I wonder if this issue doesn't have practical real-world consequences, increasing read latencies and networking costs for LOCAL_QUORUM or LOCAL_ONE reads on multi-dc clusters, when:
1. One node on one DC is rebooted
2. Possibly the workload on two DCs is slightly different, causing a permanently slightly-different hit ratios on the different DCs.

**Comment by mykaul** [comment]
@nyh - the initial description says that it does not affect local quorum?

**Comment by nyh** [comment]
You're right, sorry, I got confused. It only affects you if you use ONE, not LOCAL_ONE. And anyone sensible would use LOCAL_* on multi-DC. So false alarm. Good. Sorry for the spam.

**Comment by mykaul** [comment]
> This problem can also affect multi-**rack** setups when used with a non-topology-aware driver such as the current Alternator drivers, or when using a topology-aware driver but the request doesn't allow topology-aware routing (e.g., a scan or an unprepared statement):
> 
> Consider a client on rack A sending a request to a coordinator in rack A - choosing a coordinator in rack A to avoid a more expensive inter-rack communication. Now, consider that for one of the reasons mentioned above, this target coordinator does _not_ hold the requested data, and it needs to forward the request to a different replica. We **want** it to forward the request to the replica on the same rack (A), not to a replica on a different rack - again to reduce communication cost. And indeed, the rack-aware snitch will do this. However, if HWLB kicks in (because the hit rate is not higher than 0.95 and differs by more than 0.01 between nodes), this decision will be overridden and we'll send the request to a more-or-less random rack.
> 
> Perhaps it makes sense to increase this "0.01" constant. If there is a really significant difference between the cache hit ratio between different replicas, maybe it does make sense to override the default rack awareness and get better performance with an extra price. But we shouldn't do this for some trivial 1% benefit.

Cross AZ latencies (from https://www.bitsand.cloud/posts/cross-az-latencies/):
Top Ten Slowest Pairs
availability_zone_from	availability_zone_to	Latency (ms)
sa-east-1a	sa-east-1b	2.42
me-central-1a	me-central-1c	1.95
ap-northeast-1a	ap-northeast-1c	1.87
il-central-1a	il-central-1b	1.70
eu-north-1b	eu-north-1c	1.64
ap-northeast-1a	ap-northeast-1d	1.54
eu-west-3a	eu-west-3b	1.40
sa-east-1b	sa-east-1c	1.39
ap-northeast-2b	ap-northeast-2c	1.38
eu-west-3b	eu-west-3c	1.36
The São Paulo region has significantly higher cross-AZ latencies than all other regions - coming in not-too-far behind are the UAE, Tokyo, and Israel regions.

Top Ten Fastest Pairs
availability_zone_from	availability_zone_to	Latency (ms)
ap-northeast-3b	ap-northeast-3c	0.39
ap-southeast-4a	ap-southeast-4c	0.40
ap-south-1a	ap-south-1c	0.43
us-east-1b	us-east-1c	0.44
us-east-1a	us-east-1f	0.45
us-west-2a	us-west-2c	0.46
us-east-1d	us-east-1e	0.46
ca-central-1a	ca-central-1b	0.46
me-south-1a	me-south-1c	0.46
ap-southeast-3b	ap-southeast-3c	0.47
